self hosted models

slimsag opened the pull request:
This PR is stacked on top of all the prior work @chrsmith has done for shuffling configuration data around; it implements the new "Self hosted models" functionality.

## Configuration

Configuring a Sourcegraph instance to use self-hosted models basically involves adding some configuration like this to the site config (if you set `modelConfiguration`, you are opting in to the new system which is in early access):

```
  // Setting this field means we are opting into the new Cody model configuration system.
  "modelConfiguration": {
    // Disable use of Sourcegraph's servers for model discovery
    "sourcegraph": null,

    // Create two model providers
    "providerOverrides": [
      {
        // Our first model provider "mistral" will be a Huggingface TGI deployment which hosts our
        // mistral model for chat functionality.
        "id": "mistral",
        "displayName": "Mistral",
        "serverSideConfig": {
          "type": "huggingface-tgi",
          "endpoints": [{"url": "https://mistral.example.com/v1"}]
        },
      },
      {
        // Our second model provider "bigcode" will be a Huggingface TGI deployment which hosts our
        // bigcode/starcoder model for code completion functionality.
        "id": "bigcode",
        "displayName": "Bigcode",
        "serverSideConfig": {
          "type": "huggingface-tgi",
          "endpoints": [{"url": "http://starcoder.example.com/v1"}]
        }
      }
    ],

    // Make these two models available to Cody users
    "modelOverridesRecommendedSettings": [
      "mistral::v1::mixtral-8x7b-instruct",
      "bigcode::v1::starcoder2-7b"
    ],

    // Configure which models Cody will use by default
    "defaultModels": {
      "chat": "mistral::v1::mixtral-8x7b-instruct",
      "fastChat": "mistral::v1::mixtral-8x7b-instruct",
      "codeCompletion": "bigcode::v1::starcoder2-7b"
    }
  }
```

More advanced configurations are possible, the above is our blessed configuration for today.

## Hosting models

Another major component of this work is starting to build up recommendations around how to self-host models, which ones to use, how to configure them, etc.

For now, we've been testing with these two on a machine with dual A100s:

* Huggingface TGI (this is a Docker container for model inference, which provides an OpenAI-compatible API - and is widely popular)
* Two models:
  * Starcoder2 for code completion; specifically `bigcode/starcoder2-15b` with `eetq` 8-bit quantization.
  * Mixtral 8x7b instruct for chat; specifically `casperhansen/mixtral-instruct-awq` which uses `awq` 4-bit quantization.

This is our 'starter' configuration. Other models - specifically other starcoder 2, and mixtral instruct models - certainly work too, and higher parameter versions may of course provide better results.

Documentation for how to deploy Huggingface TGI, suggested configuration and debugging tips - coming soon.

## Advanced configuration

As part of this effort, I have added a quite extensive set of configuration knobs to to the client side model configuration (see `type ClientSideModelConfigOpenAICompatible` in this PR)

Some of these configuration options are needed for things to work at a basic level, while others (e.g. prompt customization) are not needed for basic functionality, but are very important for customers interested in self-hosting their own models.

Today, Cody clients have a number of different _autocomplete provider implementations_ which tie model-specific logic to enable autocomplete, to a provider. For example, if you use a GPT model through Azure OpenAI, the autocomplete provider for that is entirely different from what you'd get if you used a GPT model through OpenAI officially. This can lead to some subtle issues for us, and so it is worth exploring ways to have a _generalized autocomplete provider_ - and since with self-hosted models we _must_ address this problem, these configuration knobs fed to the client from the server are a pathway to doing that - initially just for self-hosted models, but in the future possibly generalized to other providers.

## Debugging facilities

Working with customers in the past to use OpenAI-compatible APIs, we've learned that debugging can be quite a pain. If you can't see what requests the Sourcegraph backend is making, and what it is getting back.. it can be quite painful to debug.

This PR implements quite extensive logging, and a `debugConnections` flag which can be turned on to enable logging of the actual request payloads and responses. This is critical when a customer is trying to add support for a new model, their own custom OpenAI API service, etc.

## Robustness

Working with customers in the past, we also learned that various parts of our backend `openai` provider were not super robust. For example, [if more than one message was present it was a fatal error](https://github.com/sourcegraph/sourcegraph/blob/main/internal/completions/client/openai/openai.go#L305), or if the SSE stream yielded `{"error"}` payloads, they would go ignored. Similarly, the SSE event stream parser we use is heavily tailored towards [the exact response structure](https://github.com/sourcegraph/sourcegraph/blob/main/internal/completions/client/openai/decoder.go#L15-L19) which OpenAI's official API returns, and is therefor quite brittle if connecting to a different SSE stream.

For this work, I have _started by forking_ our `internal/completions/client/openai` - and made a number of major improvements to it to make it more robust, handle errors better, etc.

I have also replaced the usage of a custom SSE event stream parser - which was not spec compliant and brittle - with a proper SSE event stream parser that recently popped up in the Go community: https://github.com/tmaxmax/go-sse

My intention is that after more extensive testing, this new `internal/completions/client/openaicompatible` provider will be more robust, more correct, and all around better than `internal/completions/client/openai` (and possibly the azure one) so that we can just supersede those with this new `openaicompatible` one entirely.

## Client implementation

Much of the work done in this PR is just "let the site admin configure things, and broadcast that config to the client through the new model config system."

Actually getting the clients to respect the new configuration, is a task I am tackling in future `sourcegraph/cody` PRs.

## Test plan

1. This change currently lacks any unit/regression tests, that is a major noteworthy point. I will follow-up with those in a future PR.
  * However, these changes are **incredibly** isolated, clearly only affecting customers who opt-in to this new self-hosted models configuration.
  * Most of the heavy lifting (SSE streaming, shuffling data around) is done in other well-tested codebases.
2. Manual testing has played a big role here, specifically:
  * Running a dev instance with the new configuration, actually connected to Huggingface TGI deployed on a remote server.
  * Using the new `debugConnections` mechanism (which customers would use) to directly confirm requests are going to the right places, with the right data and payloads.
  * Confirming with a new client (changes not yet landed) that autocomplete and chat functionality work.

Can we use more testing? Hell yeah, and I'm going to add it soon. Does it work quite well and have small room for error? Also yes.

## Changelog

Cody Enterprise: added a new configuration for self-hosting models. Reach out to support if you would like to use this feature as it is in early access.

chrsmith reviewed (approved):
I see what you have going on here, and have some high-level notes about the direction. But I'm happy to approve it so you can keep moving forward on this.

Though it might be nice to put some stakes in the ground and plan on unifying the "generic provider x openai" with "openaicompatible". As right now there is { azureopenai, generic/openai, openaicompatible } and that is definitely not a good position to be in.

... but we probably aren't ready to say if or when the "openai compatible" configuration knobs can be tuned in such a way that the codepath is a drop-in replacement for our `completions/client/openai` package.

WDYT?

arafatkatze reviewed (commented):
I took a long look at the PR and I didn't find any major issues but I do have questions:
1. Will you eventually be enabling Hugging face TGI so anyone can test this PR too? Right now I don't have a reference point to compare this with besides OpenAI code but I think the devil is in the details and unless I can actually get to run it and get the response back to cody and run it through a debugger it would be difficult for me to confidently say that this works great. 

2. When you say openai compatible does that mean that both the request format and response format is EXACTLY the same as openAI meaning that even if you are sending one form of request to the OpenAI endpoint the functionality you need to read the SSE events coming back from the API might be slightly different and if that were to be the case then its harder to use the same code(This is exactly why I had issues with Claude in Google Vertex because it was similar to Anthropic API but not exactly the same so I had to create separate code).  You mention that you have ran this on a few models already(in the description) so I assume the answer is a yes but I wanted to confirm again about this.

slimsag commented:
> Though it might be nice to put some stakes in the ground and plan on unifying the "generic provider x openai" with "openaicompatible". As right now there is { azureopenai, generic/openai, openaicompatible } and that is definitely not a good position to be in.
>
> ... but we probably aren't ready to say if or when the "openai compatible" configuration knobs can be tuned in such a way that the codepath is a drop-in replacement for our completions/client/openai package.

Yep, I fully agree with this.

slimsag commented:
@arafatkatze 

1. Yeah, I have my environment but am discussing with Aravind about getting a more proper environment up for others at sourcegraph to bang against this configuration and see how it works. TGI is easy to run, but the exact configuration is a bit nuanced and you need linux+nvidia GPU with lots of RAM.

2. There's lots of software that says 'we expose an OpenAI-style API to our inference' (some examples below) - the goal of this provider is to support all of these pieces of software that _claim_ to implement OpenAI API specification, with enough knobs and configuration abilities to actually support them all (many of them differ in subtle ways):
  * OpenAI's official API
  * Huggingface TGI https://huggingface.co/docs/text-generation-inference/en/messages_api
  * Ollama https://ollama.com/blog/openai-compatibility
  * AWS LISA https://github.com/awslabs/LISA?tab=readme-ov-file#background 
  * NVIDIA NIM https://www.nvidia.com/en-us/ai/
  * and many others

But for software that doesn't advertise 'we follow OpenAI-API specification', then this provider isn't intended for that.

slimsag commented on internal/modelconfig/types/configuration.go:
```diff
@@ -82,25 +146,99 @@ type SourcegraphProviderConfig struct {
-	// We currently do not have any known client-side model configuration.
-	// But later, if anything needs to be provided to Cody clients at the
-	// model-level it will go here.
+	OpenAICompatible *ClientSideModelConfigOpenAICompatible `json:"openAICompatible,omitempty"`
```
Yeah I think standard prefix approach makes more sense, let's move in that direction

slimsag commented on internal/modelconfig/types/configuration.go:
```diff
@@ -116,6 +254,34 @@ type AWSBedrockProvisionedThroughput struct {
+	//
+	// If this field is not an empty string, we treat it as an opaque string to be sent with API
+	// requests (similar to an access token) and use it for nothing else. If this field is not
+	// specified, we default to the ModelName.
```
Accepted the comment improvement. Yes, this what you think it is.

However, I think `ModelNameOverride` makes it sounds like Sourcegraph might be using the value you provide here for other purposes (e.g. detecting features based on model name, changing prompting based on model name, etc.) and I want to make it super clear that configuring this.. only impacts what model name _is sent to the API_ and we don't rely on this for anything else.

slimsag commented on schema/site.schema.json:
```diff
@@ -3191,13 +3191,9 @@
           "items": {
             "type": "string",
             "enum": [
-              "bigcode::v1::starcoder2-3b",
```
It is technically a breaking change, yes. However, since we haven't documented or shared this configuration with anyone (and, if they did find and set it on their own already.. I am pretty sure their instance would not be functional) - I feel comfortable removing it.

Unfortunately we don't have doc comments in site.schema.json - just description fields.

slimsag commented on schema/site.schema.json:
```diff
@@ -3466,6 +3462,7 @@
             "fireworks",
             "google",
             "openai",
+            "huggingface-tgi",
```
update: we chatted about this in our zoom call earlier today, and agreed this is okay because we really are trying to communicate that this is a top-level provider, well-supported, well-tested, etc.

The intention is that:

* openaicompatible provider is the 'super generic' one, that can connect to arbitrary things.
* huggingface-tgi is very specific, tested by us, etc. - it just happens to use the same implementation as `openaicompatible` provider.

slimsag commented on schema/site.schema.json:
```diff
@@ -3613,19 +3613,61 @@
-        "accessToken": {
-          "type": "string"
+        "serviceName": {
+          "description": "The name of the service being connected to, e.g. 'huggingface-tgi'. If known to Sourcegraph, it may employ various tricks to communicate more effectively with that service. Use lowcase alphanumeric and dashes only.",
```
Removed `serviceName` entirely and went with our convo here: https://github.com/sourcegraph/sourcegraph/pull/63899#discussion_r1683640074

slimsag commented:
* Renamed DebugConnections -> EnableVerboseLogs
* Entirely removed ServiceName enum and ServiceNameCustom, removed mention of nvidia-nim, etc.
* Instead, kept top-level 'huggingface-tgi' provider as that is one we have well-tested and support.
